File: services/api/routes/tsn_security.py
from __future__ import annotations

from typing import Literal

from fastapi import APIRouter, HTTPException
from fastapi.responses import FileResponse, JSONResponse
from pydantic import BaseModel, Field

from services.api.deps import tsn_security_manager
from services.api.tsn_security_service import TsnSecurityError


router = APIRouter(prefix="/tsn-security", tags=["TSN Security"])


class LoadStage(BaseModel):
    ratePps: int = Field(ge=1)
    durationSeconds: int = Field(ge=1)


class StartSecurityRun(BaseModel):
    mode: Literal["baseline", "ptp_resilience", "fuzzing", "latency_jitter", "latency_series", "latency_load", "priority_load", "priority_series"]
    target: str
    interface: str
    scopeConfirmed: bool = False
    dryRun: bool = True
    stages: list[LoadStage] = Field(default_factory=list)
    repetitions: int | None = None


def _bad_request(exc: TsnSecurityError) -> HTTPException:
    return HTTPException(status_code=400, detail=str(exc))


@router.get("/config")
def security_config():
    return tsn_security_manager.config()


@router.get("/status")
def security_status():
    return tsn_security_manager.status()


@router.get("/runs")
async def list_security_runs():
    run_ids = tsn_security_manager.run_ids()
    return JSONResponse(content={"runIds": run_ids})


@router.get("/runs/{run_id}")
async def get_security_run(run_id: str):
    try:
        return JSONResponse(content=tsn_security_manager.get_run(run_id))
    except TsnSecurityError as exc:
        raise _bad_request(exc) from exc


@router.post("/runs", status_code=202)
def start_security_run(payload: StartSecurityRun):
    try:
        return tsn_security_manager.start(payload.model_dump())
    except TsnSecurityError as exc:
        raise _bad_request(exc) from exc


@router.post("/stop")
def stop_security_run():
    return tsn_security_manager.stop_active()


@router.get("/runs/{run_id}/artifacts/{name}")
def download_security_artifact(run_id: str, name: str):
    try:
        path = tsn_security_manager.artifact(run_id, name)
    except TsnSecurityError as exc:
        raise _bad_request(exc) from exc
    return FileResponse(path, filename=path.name)

File: services/ui/dashboard/src/api/tsnSecurity.ts
import { ApiClient } from './client'

export type SecurityMode = 'baseline' | 'ptp_resilience' | 'fuzzing' | 'latency_jitter' | 'latency_series' | 'latency_load' | 'priority_load' | 'priority_series'

export type LoadStage = {
  ratePps: number
  durationSeconds: number
}

export type SecurityConfig = {
  target: string
  interfaces: string[]
  scope: string
  board3Excluded: boolean
  limits: {
    maxStageRatePps: number
    maxStageDurationSeconds: number
    maxStages: number
    fuzzMaxRatePps: number
    fuzzMaxDurationSeconds: number
    latencyMaxRatePps: number
    latencyMaxDurationSeconds: number
    latencySeriesMaxDurationSeconds: number
    latencySeriesMaxRepetitions: number
    latencyLoadMaxRatePps: number
    latencyLoadMaxDurationSeconds: number
    prioritySeriesMaxDurationSeconds: number
  }
}

export type SecurityRun = {
  runId: string
  active: boolean
  request?: {
    mode: SecurityMode
    target: string
    interface: string
    dryRun: boolean
    requestedUtc: string
  } | null
  state?: { status?: string; phase?: string; stage?: number } | null
  report?: {
    mode: SecurityMode
    status: string
    framesSent: number
    baselineReachable: boolean
    recoveryReachable: boolean
    recoverySeconds: number
    latency?: {
      available: boolean
      timestampMethod?: string | null
      matchedPackets?: number
      lossPercent?: number | null
      meanNs?: number | null
      p95Ns?: number | null
      p99Ns?: number | null
      maxNs?: number | null
      jitterStddevNs?: number | null
    } | null
    latencySeries?: {
      available: boolean
      requestedRepetitions?: number
      completedRepetitions?: number
      totalMatchedPackets?: number
      totalLossPercent?: number | null
      meanOfRunMeansNs?: number | null
      confidence95LowerNs?: number | null
      confidence95UpperNs?: number | null
      stddevOfRunMeansNs?: number | null
      outlierRepetitions?: number[]
    } | null
    latencyLoad?: {
      available: boolean
      backgroundRatePps?: number
      backgroundSentPackets?: number
      backgroundReceivedPackets?: number
      meanDeltaNs?: number | null
      p95DeltaNs?: number | null
      jitterDeltaNs?: number | null
      lossDeltaPercent?: number | null
      priorityClaim?: boolean | string
      priorityProfileApplied?: boolean
      priorityProfileRestored?: boolean
      burstCapture?: {
        available: boolean
        packets?: number
        destinationKinds?: Record<string, number>
      }
      measurementPathCapture?: {
        available: boolean
        flows?: Record<string, { egressPackets?: number; ingressPackets?: number; missingAtIngress?: number }>
      }
    } | null
    fuzzing?: {
      available: boolean
      requestedFrames?: number
      observedPackets?: number
      uniqueSequences?: number
      missingPacketsAtCapture?: number
      duplicatePackets?: number
      missingSequences?: Array<{ sequence: number; occurrences: number }>
      mutationCounts?: Record<string, number>
      reason?: string | null
    } | null
    prioritySeries?: {
      available: boolean
      requestedRepetitions?: number
      completedRepetitions?: number
      backgroundRatePps?: number
      baselineLossPercent?: number | null
      loadedLossPercent?: number | null
      meanDeltaNs?: {
        mean?: number | null
        confidence95Lower?: number | null
        confidence95Upper?: number | null
      }
      priorityProfileRestored?: boolean
    } | null
    error?: string | null
  } | null
  files?: { name: string; size: number }[]
}

export type SecurityStatus = { active: boolean; runId?: string | null; pid?: number | null }

export const securityApi = (apiBase: string) => {
  const client = new ApiClient(apiBase)
  return {
    config: () => client.get<SecurityConfig>('/api/tsn-security/config'),
    status: () => client.get<SecurityStatus>('/api/tsn-security/status'),
    runs: async () => {
      const index = await client.get<{ runIds: string[] }>('/api/tsn-security/runs')
      return Promise.all(index.runIds.map(runId => client.get<SecurityRun>(`/api/tsn-security/runs/${runId}`)))
    },
    start: (payload: {
      mode: SecurityMode
      target: string
      interface: string
      scopeConfirmed: boolean
      dryRun: boolean
      stages: LoadStage[]
      repetitions?: number
    }) => client.post<SecurityRun>('/api/tsn-security/runs', payload),
    stop: () => client.post<{ stopped: boolean; message: string }>('/api/tsn-security/stop'),
  }
}
